lux/core/sqltable.py

```diff
@@ -0,0 +1,192 @@
+
+        self._sampled = None
+        self._toggle_pandas_display = True
+        self._message = Message()
+        self._pandas_only = False
+        # Metadata
+        self._data_type = {}
+        self.unique_values = None
+        self.cardinality = None
+        self._min_max = None
+        self.pre_aggregated = None
+        self._type_override = {}
+
+        if table_name != "":
+            self.set_SQL_table(table_name)
+        warnings.formatwarning = lux.warning_format
+
+    def set_SQL_table(self, t_name):
+        # function that ties the Lux Dataframe to a SQL database table
+        if self.table_name != "":
+            warnings.warn(
+                f"\nThis dataframe is already tied to a database table. Please create a new Lux dataframe and connect it to your table '{t_name}'.",
+                stacklevel=2,
+            )
+        else:
+            self.table_name = t_name
+        import psycopg2
+
+        try:
+            lux.config.executor.compute_dataset_metadata(self)
+        except Exception as error:
+            error_str = str(error)
+            if f'relation "{t_name}" does not exist' in error_str:
+                warnings.warn(
+                    f"\nThe table '{t_name}' does not exist in your database./",
+                    stacklevel=2,
+                )
+
+    def _repr_html_(self):
+        from IPython.display import display
+        from IPython.display import clear_output
+        import ipywidgets as widgets
+
+        try:
+            if self._pandas_only:
+                display(self.display_pandas())
+                self._pandas_only = False
+            if not self.index.nlevels >= 2 or self.columns.nlevels >= 2:
+                self.maintain_metadata()
+
+                if self._intent != [] and (not hasattr(self, "_compiled") or not self._compiled):
+                    from lux.processor.Compiler import Compiler
+
+                    self.current_vis = Compiler.compile_intent(self, self._intent)
+
+            if lux.config.default_display == "lux":
+                self._toggle_pandas_display = False
+            else:
+                self._toggle_pandas_display = True
+
+            # df_to_display.maintain_recs() # compute the recommendations (TODO: This can be rendered in another thread in the background to populate self._widget)
+            self.maintain_recs()
+
+            # Observers(callback_function, listen_to_this_variable)
+            self._widget.observe(self.remove_deleted_recs, names="deletedIndices")
+            self._widget.observe(self.set_intent_on_click, names="selectedIntentIndex")
+
+            button = widgets.Button(
+                description="Toggle Data Preview/Lux",
+                layout=widgets.Layout(width="200px", top="5px"),
+            )
+            self.output = widgets.Output()
+            lux.config.executor.execute_preview(self)
+            display(button, self.output)
+
+            def on_button_clicked(b):
```

Comment:
Will the widget and display have to be re-done whenever things like `_repr_html_` change in general? For example, #291 might change the `_repr_html_` quite a bit.

Reply:
Yeah looking at #291 we would also probably have to change `_repr_html_` here to work with lazy streaming of tabs. In general most of the changes in `frame.py` from #291 can probably carry over, but we would need to adjust the default Pandas view.